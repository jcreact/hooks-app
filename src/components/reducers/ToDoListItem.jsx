import React from 'react';
import PropTypes from 'prop-types';

/**
 * Renderiza una tarea.
 * @param {Object} props - Propiedades del componente.
 * @param {number} props.index - La posici贸n de la tarea en el listado de tareas.
 * @param {{id: number, project:string, task:string, done:boolean}} props.todo - Una Tarea.
 * @param {(id: number) => void} props.removeTask - Referencia a la funci贸n para remover una tarea.
 * @param {(id: number) => void} props.handleToggle - Referencia a la funci贸n para cambiar el estado de la tarea.
 * @returns
 */
export const ToDoListItem = ({
    index,
    todo: { id, project, task, done },
    removeTask,
    handleToggle,
}) => {
    return (
        <li
            className="list-group-item d-flex justify-content-between align-items-center pointer"
            onClick={() => handleToggle(id)}
        >
            <i
                className={`bi bi-check-circle-fill ${
                    done ? 'text-success' : 'text-black-50'
                }`}
            />

            <div className="ms-2 me-auto">
                <div className="fw-bold">[{project}]</div>
                <span className={`${done && 'text-decoration-line-through'}`}>
                    {index}. {task}
                </span>
            </div>

            <div>
                <button
                    className="btn btn-outline-danger shadow-none btn-circle"
                    onClick={() => removeTask(id)}
                >
                    <i className="bi bi-trash"></i>
                </button>
            </div>
        </li>
    );
};

// Especificaci贸n de tipos de las propiedades.
ToDoListItem.propTypes = {
    index: PropTypes.number.isRequired,
    todo: PropTypes.exact({
        id: PropTypes.number.isRequired,
        project: PropTypes.string.isRequired,
        task: PropTypes.string.isRequired,
        done: PropTypes.bool.isRequired,
    }),
    removeTask: PropTypes.func.isRequired,
    handleToggle: PropTypes.func.isRequired,
};
