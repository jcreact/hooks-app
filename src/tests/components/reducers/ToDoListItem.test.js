
import { shallow } from 'enzyme';

import { ToDoListItem } from '../../../components/reducers/ToDoListItem.jsx';
import { demoTodos } from '../../fixtures/demoTodos';


describe(`Pruebas en el componente 'ToDoListItem.jsx'.`, () => {

    const id = 2;
    const index = 2;
    const todo = demoTodos.find(t => t.id === id);
    const removeTask = jest.fn();
    const handleToggle = jest.fn();

    const component = shallow(
        <ToDoListItem
            index={index}
            todo={todo}
            removeTask={removeTask}
            handleToggle={handleToggle}
        />
    );

    test(`Debe renderizar el componente ToDoListItem correctamente.`, () => {
        expect(component).toMatchSnapshot();
    });

    test(`Debe llamar la función 'removeTask' para remover una tarea.`, () => {
        component.find('button').simulate('click');
        expect(removeTask).toBeCalled();
        expect(removeTask).toBeCalledWith(id);
    });

    test(`Debe llamar la función 'handleToggle' para cambiar el estado de una tarea.`, () => {
        component.find('li').simulate('click');
        expect(handleToggle).toBeCalled();
        expect(handleToggle).toBeCalledWith(id);
    });

    test(`Debe mostrar la descripción y el proyecto de la tarea correctamente.`, () => {
        expect(component.find('span').text()).toBe(`${index + 1}. ${todo.task}`);
        expect(component.find('div.fw-bold').text()).toBe(`[${todo.project}]`);
        expect(component.find('span').hasClass('text-decoration-line-through')).toBe(todo.done);
    });

});
