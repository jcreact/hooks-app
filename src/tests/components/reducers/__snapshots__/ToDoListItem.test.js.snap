// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente 'ToDoListItem.jsx'. Debe renderizar el componente ToDoListItem correctamente. 1`] = `
<li
  className="list-group-item d-flex justify-content-between align-items-center pointer"
  onClick={[Function]}
>
  <i
    className="bi bi-check-circle-fill text-success"
  />
  <div
    className="ms-2 me-auto"
  >
    <div
      className="fw-bold"
    >
      [
      Personal
      ]
    </div>
    <span
      className="text-decoration-line-through"
    >
      3
      . 
      Tarea 2
    </span>
  </div>
  <div>
    <button
      className="btn btn-outline-danger shadow-none btn-circle"
      onClick={[Function]}
    >
      <i
        className="bi bi-trash"
      />
    </button>
  </div>
</li>
`;
